init.lua

```diff
@@ -22,14 +22,19 @@ local S = minetest.get_mod_storage()
 assert(S, "mod_storage is required")
 
 -- import warps or load
-local store = S:get("warps")
-local worldpath = minetest.get_worldpath()
-if store then
-	warps = minetest.deserialize(store)
-else
+local function firstload()
+	local store = S:get("warps")
+	local worldpath = minetest.get_worldpath()
+	if store then
+		warps = minetest.deserialize(store)
+		return
+	end
 	local fh,err = io.open(worldpath .. "/warps.txt", "r")
 	if err then
-		minetest.log("action", "[warps] loaded ")
+		-- If it doesn't exist, we've never used this mod before.
+		if not err:find("No such file or directory") then
+			minetest.log("error", "[warps] Error trying to load warps.txt: " .. err)
+		end
```

Comment:
There's barely any other issue that could occur by reading from a file. Also comparing the error message (which happens to be `nil` in my tests) is not optimal if this varies between Lua/-JIT versions or locales.

Suggestion:
```Lua
if not fh then
	return
end

while true do
	......
```

Reply:
I'm not sure about this because I just tested the following code:
```lua
fh, err = io.open("nosuchfile")
print(fh,err)
```
On these platforms:
* Windows 7 lua 5.3
* Windows 7 Luajit 2.0.5
* Debian 10 Luajit 2.1.0-beta3
* Debian 10 Lua 5.1.5

And all gave the exact same answer: `nil	nosuchfile: No such file or directory`

I then disabled mod security and tried:
```lua
//lua fh, err = io.open("nosuchfile") minetest.chat_send_all(tostring(fh)) minetest.chat_send_all(tostring(err))
```
with worldedit on these platforms:
* Windows 7 minetest 5.2.0 downloaded from minetest.net
* Debian 10 minetest 5.2.0 compiled from source with LuaJIT

and in both cases the result was the same: `nil \n nosuchfile: No such file or directory`

I would also like to quote the relevant lines of source in Luajit:
```c
lib_io.c:85-88
iof->fp = fopen(fname, mode);
  if (iof->fp == NULL)
    luaL_argerror(L, 1, lj_str_pushf(L, "%s: %s", fname, strerror(errno)));
  return iof;
```

and in Lua 5.3 from lua.org:
```c
liolib.c:263-264
  p->f = fopen(filename, mode);
  return (p->f == NULL) ? luaL_fileresult(L, 0, filename) : 1;
```
```c
lauxlib.c:245,254
_API int luaL_fileresult (lua_State *L, int stat, const char *fname) {
...
      lua_pushfstring(L, "%s: %s", fname, strerror(en));
```

Reply:
Weird. I even tested this in CLI beforehand to not falsely state something:
```
$ luajit
LuaJIT 2.1.0-beta3 -- Copyright (C) 2005-2017 Mike Pall. http://luajit.org/
JIT: ON SSE2 SSE3 SSE4.1 BMI2 fold cse dce fwd dse narrow loop abc sink fuse
> local a, b = io.open("test.txt", "r")
> print(a, b)
nil	nil
```
```
$ lua
Lua 5.1.5  Copyright (C) 1994-2012 Lua.org, PUC-Rio
> local a, b = io.open("test.txt", "r")
> print(a, b)
nil	nil
```
Well then.. in either way checking against `fh` is good enough. There's not  much that could go wrong when reading from a file.